src/rendering/MaterialParser/ConfigLoader.hh

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (C) 2023 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+// This code is adapted from https://wiki.ogre3d.org/All-purpose+script+parser
+
+#ifndef RENDERING__MATERIALPARSER__CONFIGLOADER_HH_
+#define RENDERING__MATERIALPARSER__CONFIGLOADER_HH_
+
+#include <cassert>
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#include <gz/common/Console.hh>
+
+namespace gz
+{
+namespace sim
+{
+class ConfigNode;
+
+class ConfigLoader
+{
+public:
+  static void loadMaterialFiles(ConfigLoader * c);
+
+  ConfigLoader();
+
+  ~ConfigLoader();
+
+  std::string m_fileEnding;
+
+  // For a line like
+  // entity animals/dog
+  // {
+  //    ...
+  // }
+  // The type is "entity" and the name is "animals/dog"
+  // Or if animal/dog was not there then name is ""
+  virtual ConfigNode * getConfigScript(const std::string & name);
+
+  virtual std::map<std::string, ConfigNode *> getAllConfigScripts();
+
+  virtual void parseScript(std::ifstream & stream);
+
+protected:
+  float m_LoadOrder;
+  // like "*.object"
+
+  std::map<std::string, ConfigNode *> m_scriptList;
+
+  enum Token
+  {
+    TOKEN_Text,
+    TOKEN_NewLine,
+    TOKEN_OpenBrace,
+    TOKEN_CloseBrace,
+    TOKEN_EOF,
+  };
+
+  Token tok, lastTok;
+  std::string tokVal, lastTokVal;
+
+  void _parseNodes(std::ifstream & stream, ConfigNode * parent);
+  void _nextToken(std::ifstream & stream);
+  void _skipNewLines(std::ifstream & stream);
+
+  virtual void clearScriptList();
+};
+
+class ConfigNode
+{
+public:
+  explicit ConfigNode(ConfigNode * parent,
+    const std::string & name = "untitled");
+
+  ~ConfigNode();
+
+  inline void setName(const std::string & name)
+  {
+    this->m_name = name;
+  }
+
+  inline std::string & getName()
+  {
+    return m_name;
+  }
+
+  inline void addValue(const std::string & value)
+  {
+    m_values.push_back(value);
+  }
+
+  inline void clearValues()
+  {
+    m_values.clear();
+  }
+
+  inline std::vector<std::string> & getValues()
+  {
+    return m_values;
+  }
+
+  inline void getValuesInFloat(std::vector<float> & floatValues)
+  {
+    for (const auto & str : m_values) {
+      floatValues.push_back(std::stof(str));
+    }
+  }
+
+  inline void getColorValues(std::vector<float> & colorValues)
+  {
+    getValuesInFloat(colorValues);
+    if (colorValues.size() < 3) {
```

Comment:
I noticed for specular values, the size could be 4 and we have code like this:
```c++
            values->specular.emplace(gz::math::Color(specularValues[0],
              specularValues[1], specularValues[2], specularValues[3]));
```
I have suggested in my other comment to do `colorValues.resize(3)`, but perhaps it should be `colorValues.resize(4)` to guard against out of bound access in `specularValues`.

Reply:
Good catch, adding a size parameter in the `getColorValues` function. 
Should I update the assignment of `values->ambient`  to a for loop? Since its a single assignment I thought emplace would be better otherwise I can make another function to handle `values->*` assignments instead of 3 for loops.

Reply:
I really think encapsulating all this logic in `getColorValues` and returning a `gz::math::Color` makes more sense. Then all the error checking happens in one place and we don't have to repeat for loops.

Reply:
Couldn't initialize the color in place as gz::math::Color doesn't accept vectors, so initializing based on the size (3/4).  Also moved that logic out of the header as the function got bigger to be inline.